federation/pkg/federation-controller/job/jobcontroller.go

```diff
@@ -0,0 +1,537 @@
+	batchv1 "k8s.io/kubernetes/pkg/apis/batch/v1"
+	"k8s.io/kubernetes/pkg/client/cache"
+	kubeclientset "k8s.io/kubernetes/pkg/client/clientset_generated/release_1_5"
+	"k8s.io/kubernetes/pkg/client/record"
+	"k8s.io/kubernetes/pkg/controller"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/flowcontrol"
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/util/workqueue"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+const (
+	FedJobPreferencesAnnotation = "federation.kubernetes.io/job-preferences"
+	allClustersKey              = "THE_ALL_CLUSTER_KEY"
+	UserAgentName               = "Federation-Job-Controller"
+)
+
+var (
+	jobReviewDelay          = 10 * time.Second
+	clusterAvailableDelay   = 20 * time.Second
+	clusterUnavailableDelay = 60 * time.Second
+	allJobReviewDelay       = 2 * time.Minute
+	updateTimeout           = 30 * time.Second
+	backoffInitial          = 5 * time.Second
+	backoffMax              = 1 * time.Minute
+)
+
+func parseFedJobReference(frs *batchv1.Job) (*fed.FederatedReplicaSetPreferences, error) {
+	if frs.Annotations == nil {
+		return nil, nil
+	}
+	frsPrefString, found := frs.Annotations[FedJobPreferencesAnnotation]
+	if !found {
+		return nil, nil
+	}
+	var frsPref fed.FederatedReplicaSetPreferences
+	if err := json.Unmarshal([]byte(frsPrefString), &frsPref); err != nil {
+		return nil, err
+	}
+	return &frsPref, nil
+}
+
+type JobController struct {
+	fedClient fedclientset.Interface
+
+	jobController *cache.Controller
+	jobStore      cache.Store
+
+	fedJobInformer fedutil.FederatedInformer
+
+	jobDeliverer     *fedutil.DelayingDeliverer
+	clusterDeliverer *fedutil.DelayingDeliverer
+	jobWorkQueue     workqueue.Interface
+	// For updating members of federation.
+	fedUpdater fedutil.FederatedUpdater
+
+	jobBackoff *flowcontrol.Backoff
+	// For events
+	eventRecorder record.EventRecorder
+
+	defaultPlanner *planner.Planner
+}
+
+func NewJobController(fedClient fedclientset.Interface) *JobController {
+	broadcaster := record.NewBroadcaster()
+	broadcaster.StartRecordingToSink(eventsink.NewFederatedEventSink(fedClient))
+	recorder := broadcaster.NewRecorder(api.EventSource{Component: "federated-job-controller"})
+
+	fjc := &JobController{
+		fedClient:        fedClient,
+		jobDeliverer:     fedutil.NewDelayingDeliverer(),
+		clusterDeliverer: fedutil.NewDelayingDeliverer(),
+		jobWorkQueue:     workqueue.New(),
+		jobBackoff:       flowcontrol.NewBackOff(backoffInitial, backoffMax),
+		defaultPlanner: planner.NewPlanner(&fed.FederatedReplicaSetPreferences{
+			Clusters: map[string]fed.ClusterReplicaSetPreferences{
+				"*": {Weight: 1},
+			},
+		}),
+		eventRecorder: recorder,
+	}
+
+	jobFedInformerFactory := func(cluster *fedv1.Cluster, clientset kubeclientset.Interface) (cache.Store, cache.ControllerInterface) {
+		return cache.NewInformer(
+			&cache.ListWatch{
+				ListFunc: func(options api.ListOptions) (runtime.Object, error) {
+					versionedOptions := fedutil.VersionizeV1ListOptions(options)
```

Comment:
FYI I think this is redundant now, ListFunc and Watch func take versioned params now.